compiler-core/src/language_server/tests/router.rs

```diff
@@ -0,0 +1,121 @@
+use std::time::SystemTime;
+
+use crate::{
+    Error,
+    io::FileSystemWriter,
+    language_server::{files::FileSystemProxy, tests::Action},
+    paths::ProjectPaths,
+};
+
+use super::LanguageServerTestIO;
+
+type Router = crate::language_server::router::Router<LanguageServerTestIO, LanguageServerTestIO>;
+
+#[test]
+fn recompile_after_no_changes_does_not_redownload_dependencies() {
+    let paths = ProjectPaths::new("/app".into());
+    let (io, mut router) = set_up_minimal_router(&paths);
+
+    assert_eq!(
+        compile(&mut router, &paths),
+        Ok(()),
+        "Pre-condition: Initial compile should succeed"
+    );
+
+    {
+        let mut actions = io.actions.lock().unwrap();
+        assert!(
+            actions.contains(&Action::DownloadDependencies),
+            "Expectation: Initial compile should download dependencies"
+        );
+        actions.clear();
+    }
+
+    assert_eq!(
+        compile(&mut router, &paths),
+        Ok(()),
+        "Recompile should succeed"
+    );
+
+    {
+        let actions = io.actions.lock().unwrap();
+        assert!(
+            !actions.contains(&Action::DownloadDependencies),
+            "Recompile should not re-download dependencies"
+        );
```

Comment:
Since I couldn't find any reliable way to check if a completely fresh project is set up, I'm now riding on this assumption to test things.

I'm not sure if it's the best thing to check, so I'm wondering if adding a function to `ProgressReporter` for a recreated project would make more sense? Or adding something (an id?) to `Project` to be able to check if they are newly created?

Reply:
FWIW, I tried comparing raw `Project` pointers, but at least on my machine, the new project ended up being created on the exact same address, so that's not reliable.

Reply:
I think this is good! Can always improve it later